src/transform/src/column_knowledge.rs

```diff
@@ -327,73 +327,84 @@ pub fn optimize(
-                *expr = MirScalarExpr::Literal(datum.clone(), typ.clone());
-            }
-            column_knowledge[index].clone()
-        }
-        MirScalarExpr::Literal(row, typ) => DatumKnowledge {
-            value: Some((row.clone(), typ.clone())),
-            nullable: match row {
-                Ok(row) => row.unpack_first() == Datum::Null,
-                Err(_) => false,
-            },
-        },
-        MirScalarExpr::CallNullary(_) => DatumKnowledge::default(),
-        MirScalarExpr::CallUnary { func, expr: inner } => {
-            let knowledge = optimize(inner, input_type, column_knowledge);
-            if knowledge.value.is_some() {
-                expr.reduce(input_type);
-                optimize(expr, input_type, column_knowledge)
-            } else if func == &UnaryFunc::IsNull && !knowledge.nullable {
-                *expr = MirScalarExpr::literal_ok(Datum::False, ScalarType::Bool);
-                optimize(expr, input_type, column_knowledge)
-            } else {
-                DatumKnowledge::default()
-            }
-        }
-        MirScalarExpr::CallBinary {
-            func: _,
-            expr1,
-            expr2,
-        } => {
-            let knowledge1 = optimize(expr1, input_type, column_knowledge);
-            let knowledge2 = optimize(expr2, input_type, column_knowledge);
-            if knowledge1.value.is_some() && knowledge2.value.is_some() {
-                expr.reduce(input_type);
-                optimize(expr, input_type, column_knowledge)
-            } else {
-                DatumKnowledge::default()
-            }
-        }
-        MirScalarExpr::CallVariadic { func: _, exprs } => {
-            let mut knows = Vec::new();
-            for expr in exprs.iter_mut() {
-                knows.push(optimize(expr, input_type, column_knowledge));
-            }
+    // Storage for `DatumKnowledge` being propagated up through the
+    // `MirScalarExpr`. When a node is visited, pop off as many `DatumKnowledge`
+    // as the number of children the node has, and then push the
+    // `DatumKnowledge` corresponding to the node back onto the stack.
+    // Post-order traversal means that if a node has `n` children, the top `n`
+    // `DatumKnowledge` in the stack are the `DatumKnowledge` corresponding to
+    // the children.
+    let mut knowledge_stack = Vec::<DatumKnowledge>::new();
 
-            if knows.iter().all(|k| k.value.is_some()) {
-                expr.reduce(input_type);
-                optimize(expr, input_type, column_knowledge)
+    expr.visit_custom_mut(
+        &mut |e| {
+            // We should not eagerly memoize `if` branches that might not be taken.
+            // TODO: Memoize expressions in the intersection of `then` and `els`.
+            if let MirScalarExpr::If { then, els, .. } = e {
+                Some(vec![then, els])
             } else {
-                DatumKnowledge::default()
+                None
```

Comment:
Reading this, it is slightly weird to have `None` mean "all". I'm not sure there is a problem, but I scratched my head for several seconds before I figured it out.

Reply:
Thinking about it more, I do like that the "default" behavior is easy and cheap to do. Maybe just weird that `None` is the wrong word. If enums were easy to use, `All` vs `Some` would clearly be great, but that is probably horrible wrt importing enums and all of that. =/